### PR TITLE
ExtraPortMapping Validation

### LINKED ISSUE #3301: A unique hostPort is not generated when using extraPortMappings without specifying hostPort.
**What happened**:

Creating a kind cluster without any `hostPort` defined in the `extraPortMappings` should use random but unique `hostPort`. However it seems like it's using the same and so fails before cluster creation during validation. This was/is working as expected with kind v0.18.0

There is a workaround if creating only 1 cluster, which is to manually add `hostPort` and select a unique port. However this does not scale when we need to run multiple kind cluster. (eg for demonstrating or testing a multi cluster to replacate a multi region behaviour etc).

**What you expected to happen**:
The kind cluster to be created successfully

**How to reproduce it (as _minimally_ and precisely as possible)**:
Given the below config file and creating a cluster will fail with the folloing error.
```
ERROR: failed to create cluster: invalid configuration for node 1: invalid portMapping: port mapping with same listen address, port and protocol already configured: <nil>:0/
```

command run: `kind create cluster --config config.yaml`

```yaml
# config.yaml

kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
nodes:
  - role: control-plane
  - role: worker
    extraPortMappings:
      - containerPort: 8080
      - containerPort: 8443
```

**Anything else we need to know?**:
This error only occurs after #3175 so all version from v0.19.0 on wards.
v0.18.0 continues to work as expected.

**Environment:**

- kind version: (use `kind version`): v0.20.0
- Runtime info: (use `docker info` or `podman info`): Docker engine 24.0.2
- OS (e.g. from `/etc/os-release`): OSX 13.4
- Kubernetes version: (use `kubectl version`): 1.24.15
- Any proxies or other special environment settings?: nil

### PR DESCRIPTION
**Pull request by aroradaman**
Using random host ports for validating extra port mapping defined without host ports  (default behaviour for provider docker and podman)

Fixes #3301

### CODE REVIEW

File: pkg/internal/apis/config/validate.go [hunk 1]
```diff
@@ -151,7 +152,39 @@ func validatePortMappings(portMappings []PortMapping) error {
...
 	for _, portMapping := range portMappings {
+		// using default listen address
+		if portMapping.ListenAddress == "" {
+			portMapping.ListenAddress = wildcardAddrIPv4.String()
+		}
+
+		// using random hostport
+		// this is just for validation purpose, actual random host port will be assigned by pkg/cluster/internal/providers
+		if portMapping.HostPort == 0 {
+			portMapping.HostPort = getRandomHostPort()
```
Comment (aojea) [inline pkg/internal/apis/config/validate.go:180]: I don't think we should fake the random port, it looks like if this value is empty we have to assume there will not be conflict and just skip its validation

Comment (aroradaman) [reply pkg/internal/apis/config/validate.go]: makes more sense :)

File: pkg/internal/apis/config/validate_test.go [hunk 2]
```diff
@@ -432,45 +432,66 @@ func TestPortValidate(t *testing.T) {
...
+		},
+		{
+			testName: "duplicate container ports",
+			portMappings: []PortMapping{
+				newPortMapping("", 0, 1000, ""),
+				newPortMapping("", 0, 2000, ""),
+				newPortMapping("", 0, 3000, ""),
+				newPortMapping("", 0, 3000, ""),
+			},
+			expectErr: "",
```
Comment (aojea) [inline pkg/internal/apis/config/validate_test.go:466]: both testcases are the same, aren't they?

Comment (aroradaman) [reply pkg/internal/apis/config/validate_test.go:466]: the second one has a duplicate containerPort, which is allowed by both kind and docker.

Comment (aojea) [reply pkg/internal/apis/config/validate_test.go:466]: oh, heh, I started at it for one minute and didn't notice it XD

File: pkg/internal/apis/config/validate.go [hunk 3]
```diff
@@ -152,11 +153,26 @@ func validatePortMappings(portMappings []PortMapping) error {
...
+
+		// using default listen address
+		if portMapping.ListenAddress == "" {
+			portMapping.ListenAddress = wildcardAddrIPv4.String()
+		}
+
+		// using default protool
+		if portMapping.Protocol == "" {
+			portMapping.Protocol = PortMappingProtocolTCP
+		}
```
[the remark this answers is not in the document]
Comment (aojea) [reply pkg/internal/apis/config/validate.go:169]: point is we should default things in defaulting stage, not here in validation

[the remark this answers is not in the document]
Comment (aroradaman) [reply pkg/internal/apis/config/validate.go:169]: should I remove defaults from `pkg/cluster/internal/providers/<provider>/provision.go` then and add it as part of **SetDefaultsCluster** in default stage .. defaulting with respect to provider ?

https://github.com/kubernetes-sigs/kind/blob/3610f606516ccaa88aa098465d8c13af70937050/pkg/cluster/internal/providers/docker/provision.go#L363-L377

https://github.com/kubernetes-sigs/kind/blob/3610f606516ccaa88aa098465d8c13af70937050/pkg/cluster/internal/providers/podman/provision.go#L376-L390

File: pkg/apis/config/v1alpha4/default.go [hunk 4]
```diff
@@ -79,12 +79,26 @@ func SetDefaultsCluster(obj *Cluster) {
...
+				obj.ExtraPortMappings[i].ListenAddress = defaults.ExtraPortMappingListenAddressIPv6
+			} else {
+				obj.ExtraPortMappings[i].ListenAddress = defaults.ExtraPortMappingListenAddressIPv4
+			}
+		}
+
+		if string(obj.ExtraPortMappings[i].Protocol) == "" {
+			obj.ExtraPortMappings[i].Protocol = PortMappingProtocolTCP
+		}
+	}
```
Comment (aojea) [inline pkg/apis/config/v1alpha4/default.go:103]: @BenTheElder we just pre-default here, I think is ok, but it will better to have your thoughts

**Comment by aojea** [comment]
/lgtm
/assign @BenTheElder 

for final approval

**Comment by aojea** [comment]
/lgtm cancel

don't we need to modify `./pkg/internal/apis/config/default.go` too?

File: pkg/apis/config/v1alpha4/default.go (hunk 4 above)
Comment (aroradaman) [reply pkg/apis/config/v1alpha4/default.go:103]: I will also remove the defaulting piece from docker/podman providers  if we finalize this approach.

**Comment by aroradaman** [comment]
@aojea  Is `./pkg/internal/apis/config/default.go` generated by code-generator ?
If yes, can you point me to documentation for the same ?

File: pkg/apis/config/defaults/networking.go [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+const ServiceSubnetIPv4 = "10.96.0.0/16"
+
+// ServiceSubnetIPv6 is the default IPv6 subnet for the Networking.PodSubnet field
+const ServiceSubnetIPv6 = "fd00:10:96::/112"
+
+// ServiceSubnetDualStack is the default DualStack subnet for the Networking.ServiceSubnet field
+const ServiceSubnetDualStack = "10.96.0.0/16,fd00:10:96::/112"
+
+// ExtraPortMappingListenAddressIPv4 is the default IPv4/DualStack listen address for ExtraPortMappings
+const ExtraPortMappingListenAddressIPv4 = "0.0.0.0"
```
Comment (BenTheElder) [inline pkg/apis/config/defaults/networking.go:52]: did we really need to make all of these public variables?

even if the values are documented, that doesn't mean we've committed to thoughtful names yet and that the specifics of how traffic works on a future provider will be the same.

**Review by BenTheElder — COMMENTED** [review]
I've been out sick and I'm going to need some more time to think about how to better structure this, but I don't think exporting all of these values is the answer.

Thanks for working on this problem!

**Comment by prahaladramji** [comment]
> I've been out sick and I'm going to need some more time to think about how to better structure this, but I don't think exporting all of these values is the answer.
> 
> Thanks for working on this problem!

Just following up to find out what's the current thoughts on this PR, I'm looking for a resolution to https://github.com/kubernetes-sigs/kind/issues/3301.

@BenTheElder Is the feedback more around the implementation details of this PR or is the above mentioned issue the desired behaviour of kind moving forward post v0.18.0?

**Comment by BenTheElder** [comment]
I was hoping for a response to the existing comments before reviewing further.

File: pkg/apis/config/v1alpha4/default.go (hunk 4 above)
Comment (BenTheElder) [reply pkg/apis/config/v1alpha4/default.go:102]: I guess this is OK, we can define the behavior based on docker. Not the first time kubernetes has done that.

File: pkg/apis/config/defaults/networking.go (hunk 5 above)
Comment (BenTheElder) [reply pkg/apis/config/defaults/networking.go:52]: Still not thrilled about this aspect, we're exporting a lot of internal choices here that should be versioned with the API and we're doing so as a ton of unversioned globals ...

Comment (BenTheElder) [reply pkg/apis/config/defaults/networking.go:52]: This change does not seem necessary for the purpose of this PR?

Comment (aroradaman) [reply pkg/apis/config/defaults/networking.go]: moved back to versioned defaulting.

**Comment by aroradaman** [comment]
@prahaladramji I don't think retest would help here, and also the error is not related to this PR. 

@BenTheElder do we still need `pull-kind-e2e-kubernetes-1-24` given that it has already reached EOL. I also see `pull-kind-e2e-kubernetes-1-28` to be missing.

**Comment by aroradaman** [comment]
ping @BenTheElder

**Comment by prahaladramji** [comment]
The 1 failing check seemes to be related to vagrant as part of the ci step and not related to this change.

A general look at that error seems like it could just be flakey vagrant step (potentially a retry may work). But if not, I'm happy to try and help look at fixing that step.

**Comment by aroradaman** [comment]
/close in favor of https://github.com/kubernetes-sigs/kind/pull/3513
cc. @prahaladramji

**Pull request closed without merging**